### PR TITLE
TYP: avoid inherit_names for DatetimeIndexOpsMixin

### PR DESCRIPTION
**Pull request by twoertwein**
Similar to #36742
 
xref #32100

### CODE REVIEW

**Comment by twoertwein** [comment]
Over multiple PRs, my goal is to remove `inherit_names` to make static type checkers aware of those methods.

The alternative is to keep `inherit_names` but add `method_name: type_annotation`. Writing those annotations is tricky for properties.

Adding little helper methods as done in #36742 and here has the advantage that we can easily express the type of properties, it is more difficult for the implementation and the annotations to get out of sync, and it might also be more readable. Mypy doesn't seem to support properties with decorators but it seems that it just ignores them, which should be fine in most/all cases (`@doc(...)` doesn't change the function signature)

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -214,7 +248,8 @@ def _summary(self, name=None) -> str:
     def _can_partial_date_slice(self, reso: Resolution) -> bool:
         # e.g. test_getitem_setitem_periodindex
         # History of conversation GH#3452, GH#3931, GH#2369, GH#14826
-        return reso > self._resolution_obj
+        # error: Unsupported operand types for < ("str" and "Resolution")
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimelike.py:251]: it thinks that reso is a str?

Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:251]: must have copied the wrong error message. The "correct" error is

> error: Unsupported left operand type for > ("Resolution")  [operator] 
> note: Right operand is of type "Optional[Resolution]"

Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:251]: I added a breakpoint, both of them are Resolution objects

**Review by mroeschke — APPROVED** [review]
LGTM cc @jbrockmendel

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -392,7 +391,8 @@ def is_full(self) -> bool:
         if not self.is_monotonic_increasing:
             raise ValueError("Index is not monotonic")
         values = self.asi8
-        return ((values[1:] - values[:-1]) < 2).all()
+        # error: Incompatible return value type (got "bool_", expected "bool")
```
Comment (jbrockmendel) [inline pandas/core/indexes/period.py:394]: worth wrapping in `bool(...)`?

File: pandas/core/indexes/datetimelike.py [hunk 3]
```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
...
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.freqstr)
+    def freqstr(self) -> str | None:
+        return self._data.freqstr
+
+    # error: Decorated property not supported
+    @cache_readonly  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.inferred_freq)
+    def inferred_freq(self) -> str | None:
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimelike.py:116]: i think this is only for DTI/TDI, not PI?

File: pandas/core/indexes/datetimelike.py [hunk 4]
```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
...
+        return self._data.inferred_freq
+
+    @cache_readonly
+    def _resolution_obj(self) -> Resolution | None:
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimelike.py:120]: is None possible?

File: pandas/core/indexes/datetimelike.py [hunk 5]
```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
...
+        return self._data.asi8
+
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.freq)
+    def freq(self) -> BaseOffset | None:
+        return self._data.freq
+
+    @freq.setter
+    def freq(self, value) -> None:
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimelike.py:103]: do we allow setting?

File: pandas/core/arrays/timedeltas.py [hunk 6]
```diff
@@ -926,7 +927,9 @@ def sequence_to_td64ns(
     data = np.array(data, copy=copy)
 
     assert data.dtype == "m8[ns]", data
-    return data, inferred_freq
+    # error: Incompatible return value type (got "Tuple[Any, Optional[BaseOffset]]",
+    # expected "Tuple[ndarray[Any, Any], Optional[Tick]]")
```
Comment (jbrockmendel) [inline pandas/core/arrays/timedeltas.py:931]: looks like it doesn't recognize that `data` is an ndarray here?

File: pandas/core/arrays/timedeltas.py [hunk 7]
```diff
@@ -446,7 +446,8 @@ def __truediv__(self, other):
             freq = None
             if self.freq is not None:
                 # Tick division is not implemented, so operate on Timedelta
-                freq = self.freq.delta / other
+                # error: "BaseOffset" has no attribute "delta"
```
Comment (jbrockmendel) [inline pandas/core/arrays/timedeltas.py:449]: TimedeltaArray.freq should be `Tick|None`

File: pandas/core/arrays/datetimelike.py [hunk 8]
```diff
@@ -1526,8 +1539,9 @@ def __sub__(self, other):
         elif is_integer_dtype(other_dtype):
             if not is_period_dtype(self.dtype):
                 raise integer_op_not_supported(self)
+            # error: Item "None" of "Optional[BaseOffset]" has no attribute "n"
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1542]: PeriodArray.freq should be non-None

**Comment by jbrockmendel** [comment]
Not a hill I plan to die on, but it isn't obvious to me that this is worthwhile.  we implemented inherit_names in the first place for boilerplate-reduction, which is still a good goal.

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:120]: `DatetimeLikeArrayMixin._resolution_obj` can theoretically be `None` but that is never the case when running the tests!

I could make sense to simplify `DatetimeLikeArrayMixin._resolution_obj` https://github.com/pandas-dev/pandas/blob/926b9ceff10d9b7a957811f0a4de3167332196de/pandas/core/arrays/datetimelike.py#L968 to

```py
    @property
    def _resolution_obj(self) -> Resolution:
        freqstr = self.freqstr
        return Resolution.get_reso_from_freqstr(freqstr)  # the KeyError is never triggered in the tests
```

File: pandas/core/indexes/datetimelike.py (hunk 5 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:103]: It works at least on main `pd.TimedeltaIndex([]).freq = None` but I don't know whether that is meant to be the case.

Happy to remove the setter (seems like that would be something that is not possible with `inherit_names`: it would copy both the getter and setter),

File: pandas/core/indexes/datetimelike.py (hunk 3 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:116]: On 1.4.3 (and on this PR) it errors for PeriodIndex:

```py
>>> pd.PeriodIndex(year=[2000, 2002], quarter=[1, 3]).inferred_freq
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/properties.pyx", line 37, in pandas._libs.properties.CachedProperty.__get__
  File "pandas/core/indexes/extension.py", line 59, in cached
    return getattr(self._data, name)
  File "pandas/core/arrays/datetimelike.py", line 917, in inferred_freq
    return frequencies.infer_freq(self)
  File "pandas/tseries/frequencies.py", line 183, in infer_freq
    raise TypeError(
TypeError: PeriodIndex given. Check the `freq` attribute instead of using infer_freq.
```

Could add a `NoReturn`-annotation to `freqstr` in `PeriodIndex`.

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:120]: mh, in test pandas/tests/arrays/test_datetimes.py(82)test_fields(), `self.freq` is None but `_resolution_obj` is not called during this test but it would return `None` in this case. I came across this when testing whether `freqstr` can be `None` (and in this case it is `None`): if `freqstr` can be `None`, `_resolution_obj` can be `None`.

One argument for removing `None` from at least `_resolution_obj` (but not `freqstr`) is that places that call it, do not check for None and would anyways error there

File: pandas/core/indexes/datetimelike.py (hunk 5 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:103]: It cannot be set for `PeriodArray` but other arrays (and index) types support it.

pandas/tests/arrays/categorical/test_constructors.py has two tests that set freq to None (but the tests also pass when removing those lines).

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: Yah, looks like it is only relevant for TDA, and we could re-use the logic (minus self.tz) from DTA._resolution_obj or that

File: pandas/core/indexes/datetimelike.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py:116]: yah, so better to just move this to DatetimeTimedeltaMixin?

Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py]: Thank you! I moved it.

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py]: Sorry, I'm not following: do you suggest moving `_resolution_obj` (or `freqstr`) to a different class?

File: pandas/core/indexes/datetimelike.py (hunk 5 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py]: I moved it to `DatetimeTimedeltaMixin`

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:110]: `DatetimeLikeArrayMixin._resolution_obj` is only ever called for `PeriodArray` in the tests (other arrays might implement their own version?) and `DatetimeIndexOpsMixin._resolution_obj` is never called at least in the tests (I guess sub-classes implement it).

Cannot make `DatetimeIndexOpsMixin._resolution_obj` abstract as `DatetimeIndex` and `TimedeltaIndex` do not "implement" it (`DatetimeIndex` uses `inherit_names` to replace it but it didn't see anything for `TimedeltaIndex`).

Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py:110]: I would be fine leaving it as `Resolution | None` for now.

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py:110]: if we don't need TDA._resolution_obj to exist, then we can just move it.  otherwise, i think something like

```
@property
def _resolution_obj(self) -> Resolution:
     tz = getattr(self, "tz", None)  # i.e. None for TimedeltaArray
    return get_resolution(self.asi8, tz, reso=self._reso)
```

works for DTA/TDA, and the current DatetimeLikeArrayMixin._resolution_obj becomes PeriodArray._resolution_obj

Comment (twoertwein) [reply pandas/core/indexes/datetimelike.py]: Is `TimedeltaIndex` supposed to have `_resolution_obj` (and `resolution`):

```
>>> pd.TimedeltaIndex([1])._resolution_obj  # None
>>> pd.TimedeltaIndex([1]).resolution
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/properties.pyx", line 37, in pandas._libs.properties.CachedProperty.__get__
  File "python3.10/site-packages/pandas/core/indexes/extension.py", line 59, in cached
    return getattr(self._data, name)
  File "python3.10/site-packages/pandas/core/arrays/datetimelike.py", line 937, in resolution
    return self._resolution_obj.attrname  # type: ignore[union-attr]
AttributeError: 'NoneType' object has no attribute 'attrname'
```

If it is not supposed to have that, a good solution (for at least for the index classes) would be to make `DatetimeIndexOpsMixin._resolution_obj -> Resolution` abstract (all sub-classes except TimedeltaIndex implement it and return `Resolution`).

**Comment by twoertwein** [comment]
I think I addressed all comments but now we also have some small implementation changes. If you could please have another look @jbrockmendel  @mroeschke

**Comment by mroeschke** [comment]
Generally looks okay to me. I don't have a strong opinion in which direction `inherit_names` should go (but generally I don't mind some boilerplating as it makes the code easier to grok IMO)

File: pandas/core/arrays/datetimelike.py [hunk 9]
```diff
@@ -552,7 +552,12 @@ def _concat_same_type(
 
             if obj.freq is not None and all(x.freq == obj.freq for x in to_concat):
                 pairs = zip(to_concat[:-1], to_concat[1:])
-                if all(pair[0][-1] + obj.freq == pair[1][0] for pair in pairs):
+                # error: No overload variant of "__radd__" of "BaseOffset" matches
+                # argument type "NaTType"
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:556]: is this something that should be fixed in the BaseOffset .pyi file?  (OK if you want to leave that as out of scope)

File: pandas/core/arrays/datetimelike.py [hunk 10]
```diff
@@ -1220,7 +1225,10 @@ def _sub_period(self, other: Period) -> npt.NDArray[np.object_]:
         new_i8_data = checked_add_with_arr(
             self.asi8, -other.ordinal, arr_mask=self._isnan
         )
-        new_data = np.array([self.freq.base * x for x in new_i8_data])
+        # error: Item "None" of "Optional[BaseOffset]" has no attribute "base"
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1228]: can do `self = cast("PeriodArray", self)` above

File: pandas/core/arrays/datetimelike.py [hunk 11]
```diff
@@ -1427,8 +1435,10 @@ def _time_shift(
         if self.freq is None:
             raise NullFrequencyError("Cannot shift with no freq")
 
-        start = self[0] + periods * self.freq
-        end = self[-1] + periods * self.freq
+        # error: No overload variant of "__radd__" of "BaseOffset" matches
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1438]: not sure if there's any way to annotate this, but the fact that we excluded `self.freq is None` above means self[0] and self[-1] can't be NaT

File: pandas/core/arrays/datetimelike.py [hunk 12]
```diff
@@ -1456,8 +1466,9 @@ def __add__(self, other):
             # as is_integer returns True for these
             if not is_period_dtype(self.dtype):
                 raise integer_op_not_supported(self)
+            # error: Item "None" of "Optional[BaseOffset]" has no attribute "n"
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:1469]: i think need to do the same cast on the self in self.freq.n?

File: pandas/core/arrays/timedeltas.py [hunk 13]
```diff
@@ -143,6 +144,23 @@ def _box_func(self, x: np.timedelta64) -> Timedelta | NaTType:
             return NaT
         return Timedelta._from_value_and_reso(y, reso=self._reso)
 
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(dtl.DatetimeLikeArrayMixin.freq)
+    def freq(self) -> Tick:
```
Comment (jbrockmendel) [inline pandas/core/arrays/timedeltas.py:150]: TDA.freq _can_ be None.  its only PeriodArray.freq that can't be None

File: pandas/core/indexes/period.py [hunk 14]
```diff
@@ -393,7 +392,7 @@ def is_full(self) -> bool:
         if not self.is_monotonic_increasing:
             raise ValueError("Index is not monotonic")
         values = self.asi8
-        return ((values[1:] - values[:-1]) < 2).all()
+        return ((values[1:] - values[:-1]) < 2).all().item()
```
Comment (jbrockmendel) [inline pandas/core/indexes/period.py:395]: why is this needed?  seems like the kind of thing that might be fragile across numpy versions

Comment (twoertwein) [reply pandas/core/indexes/period.py:395]: The function is annotated as returning bool, but it currently return numpy.bool_ (which does not seem to inherit from bool):

```py
>>> type((np.random.rand(3) > 0.5).all())
<class 'numpy.bool_'>
>>> isinstance((np.random.rand(3) > 0.5).all(), bool)
False
```

File: pandas/core/arrays/datetimelike.py (hunk 9 above)
Comment (twoertwein) [reply pandas/core/arrays/datetimelike.py]: I added `NaTType` for just `__radd__`. Would need to add it to many more oprators.

File: pandas/core/arrays/timedeltas.py (hunk 13 above)
Comment (twoertwein) [reply pandas/core/arrays/timedeltas.py]: Changed :)

File: pandas/core/arrays/datetimelike.py (hunk 11 above)
Comment (twoertwein) [reply pandas/core/arrays/datetimelike.py]: this is now anyways covered by the new NaTType overload in offsets.pyi :)

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by twoertwein** [comment]
> Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

Opened #49804
